command/server_util.go

```diff
@@ -14,63 +13,78 @@ var (
+		// We have the same barrier type and the unwrap seal is nil so we're not
+		// migrating from same to same, IOW we assume it's not a migration
+		if existBarrierSealConfig.Type == barrierSeal.BarrierType() {
 			return nil
 		}
 
-		if existBarrierSealConfig.Type != vaultseal.Shamir && existRecoverySealConfig == nil {
-			return errors.New(`Recovery seal configuration not found for existing seal`)
+		// If we're not coming from Shamir, and the existing type doesn't match
+		// the barrier type, we need both the migration seal and the new seal
+		if existBarrierSealConfig.Type != vaultseal.Shamir && barrierSeal.BarrierType() != vaultseal.Shamir {
+			return errors.New(`Trying to migrate from auto-seal to auto-seal but no "disabled" seal stanza found`)
+		}
+	} else {
+		if unwrapSeal.BarrierType() == vaultseal.Shamir {
+			return errors.New("Shamir seals cannot be set disabled (they should simply not be set)")
 		}
+	}
 
-		switch existBarrierSealConfig.Type {
-		case vaultseal.Shamir:
-			// The value reflected in config is what we're going to
-			existSeal = vault.NewDefaultSeal()
-			existSeal.SetCore(core)
-			newSeal = seal
-			newBarrierSealConfig := &vault.SealConfig{
-				Type:            newSeal.BarrierType(),
-				SecretShares:    1,
-				SecretThreshold: 1,
-				StoredShares:    1,
-			}
-			newSeal.SetCachedBarrierConfig(newBarrierSealConfig)
-			newSeal.SetCachedRecoveryConfig(existBarrierSealConfig)
+	var existSeal vault.Seal
+	var newSeal vault.Seal
+
+	if existBarrierSealConfig.Type == barrierSeal.BarrierType() {
+		// In this case our migration seal is set so we are using it
+		// (potentially) for unwrapping. Set it on core for that purpose then
+		// exit.
+		core.SetSealsForMigration(nil, nil, unwrapSeal)
+		return nil
+	}
 
-		default:
-			if onEnterprise {
-				return errors.New("Migrating from autoseal to Shamir seal is not supported on Vault Enterprise")
-			}
+	if existBarrierSealConfig.Type != vaultseal.Shamir && existRecoverySealConfig == nil {
+		return errors.New(`Recovery seal configuration not found for existing seal`)
+	}
 
-			// The disabled value reflected in config is what we're going from
-			existSeal = coreConfig.Seal
-			newSeal = vault.NewDefaultSeal()
-			newSeal.SetCore(core)
-			newSeal.SetCachedBarrierConfig(existRecoverySealConfig)
+	switch existBarrierSealConfig.Type {
+	case vaultseal.Shamir:
+		// The value reflected in config is what we're going to
+		existSeal = vault.NewDefaultSeal()
+		newSeal = barrierSeal
+		newBarrierSealConfig := &vault.SealConfig{
+			Type:            newSeal.BarrierType(),
+			SecretShares:    1,
+			SecretThreshold: 1,
+			StoredShares:    1,
 		}
+		newSeal.SetCachedBarrierConfig(newBarrierSealConfig)
+		newSeal.SetCachedRecoveryConfig(existBarrierSealConfig)
 
-		core.SetSealsForMigration(existSeal, newSeal)
+	default:
+		if onEnterprise {
+			return errors.New("Migrating from autoseal to Shamir seal is not currently supported on Vault Enterprise")
```

Comment:
i May have not followed the logic correctly, but wouldn’t we get into this block if we are going from auto seal to auto seal?

Reply:
Very good catch. This should be if on enterprise *and* the new seal is shamir. I'd be interested in a follow up conversation around whether this is still necessary/ramifications.